src/main/scala/rules/MagicWandSupporter.scala

```diff
@@ -331,68 +334,118 @@ object magicWandSupporter extends SymbolicExecutionRules {
+   * @param pve Partial Verification Error that is used to report errors.
+   * @param v Verifier instance.
+   * @param Q Continuation-style function that is called with the resulting state and the verification result.
+   * @return Result of the overall verification process.
+   */
   def applyWand(s: State,
                 wand: ast.MagicWand,
                 pve: PartialVerificationError,
                 v: Verifier)
                (Q: (State, Verifier) => VerificationResult)
                : VerificationResult = {
-        consume(s, wand, pve, v)((s1, snap, v1) => {
-          val wandSnap = MagicWandSnapshot(snap)
-          consume(s1, wand.left, pve, v1)((s2, snap, v2) => {
-            /* It is assumed that snap and wandSnap.abstractLhs are structurally the same.
-             * Since a wand can only be applied once, equating the two snapshots is sound.
-             */
-            assert(snap.sort == sorts.Snap, s"expected snapshot but found: $snap")
-            v2.decider.assume(snap === wandSnap.abstractLhs)
-            val s3 = s2.copy(oldHeaps = s1.oldHeaps + (Verifier.MAGIC_WAND_LHS_STATE_LABEL -> magicWandSupporter.getEvalHeap(s1)))
-            produce(s3.copy(conservingSnapshotGeneration = true), toSf(wandSnap.rhsSnapshot), wand.right, pve, v2)((s4, v3) => {
-              val s5 = s4.copy(g = s1.g, conservingSnapshotGeneration = s3.conservingSnapshotGeneration)
-              val s6 = v3.stateConsolidator(s5).consolidate(s5, v3).copy(oldHeaps = s1.oldHeaps)
-              Q(s6, v3)})})})}
+    // Consume the magic wand instance "A --* B".
+    consume(s, wand, pve, v)((s1, snapWand, v1) => {
+      // Consume the wand's LHS "A".
+      consume(s1, wand.left, pve, v1)((s2, snapLhs, v2) => {
+        /* It is assumed that snap and MagicWandSnapshot.abstractLhs are structurally the same.
+         * Equating the two snapshots is sound iff a wand is applied only once.
+         * The old solution in this case did use this assumption:
+         * v2.decider.assume(snap === snapWand.abstractLhs)
+         */
+        assert(snapLhs.sort == sorts.Snap, s"expected snapshot but found: $snapLhs")
+
+        // Create copy of the state with a new labelled heap (i.e. `oldHeaps`) called "lhs".
+        val s3 = s2.copy(oldHeaps = s1.oldHeaps + (Verifier.MAGIC_WAND_LHS_STATE_LABEL -> this.getEvalHeap(s1)))
+
+        // If the snapWand is a (wrapped) MagicWandSnapshot then lookup the snapshot of the right-hand side by applying snapLhs.
+        val magicWandSnapshotLookup = snapWand match {
+          case snapshot: MagicWandSnapSingleton =>
+            v2.decider.assume(snapLhs === snapshot.abstractLhs)
+            snapshot.rhsSnapshot
+          case snapshot: MagicWandSnapFunction => snapshot.applyToWandMap(snapLhs)
+          case SortWrapper(snapshot: MagicWandSnapFunction, _) => snapshot.applyToWandMap(snapLhs)
+          // Fallback solution for quantified magic wands
+          case predicateLookup: PredicateLookup =>
```

Comment:
Why could there be a PredicateLookup here? The code below seems to only create MWSSingletons and MWSFunctions.

Reply:
If the wand is used as a quantified resource, we get a `PredicateLookup` when consuming that magic wand. An example for this is the following program that I've extracted from the test file [QPWands.vpr](https://github.com/viperproject/silver/blob/8e84e064d890fc28ed20678a9740ccc5760d2d91/src/test/resources/wands/new_syntax/QPWands.vpr).

```viper
field f: Int

method test14(xs: Seq[Ref], y: Ref)
    requires forall x: Ref :: x in xs ==> acc(x.f)
    requires |xs| >= 1
    requires acc(y.f)
{
    xs[0].f := 0
    var some: Ref := xs[0]

    package acc(y.f) --* acc(some.f) && acc(y.f)

    // Makes a quantified magic wand out of it
    var completed: Seq[Ref] := Seq(some)
    assert forall x: Ref :: x in completed ==> acc(y.f) --* acc(x.f) && acc(y.f)

    assert applying (acc(y.f) --* acc(some.f) && acc(y.f)) in some.f == 0
}
```

Without the two lines after the comment, we have a normal magic wand and we get a `MagicWandSnapshot` here. With those two lines we have `PredicateLookup` here. Because I did not find out how to handle that case with a `MagicWandSnapFunction` I chose to use a fallback that does the same as the initial version.

Reply:
Right, the PR only changes the encoding of non-quantified wands, makes sense.
I guess that means for quantified wands, we still always have the same encoding that's potentially unsound when it's applied multiple times, right? Is that intentional or is it an accidental omission?

Reply:
This was intentional. I've tried to do it, but after one day of debugging a 2000-5000 line SMT2 file I decided to move on to other more interesting test cases for now. And yes, the current version with `PredicateLookup` is potentially unsound. I've written a test case that illustrates this behavior [here](https://github.com/manud99/silver/blob/e0c11acdefdbbd9dd4abac445dfe041f2033d7c4/src/test/resources/wands/snap_functions/test12.vpr). It verifies unsound using both versions of Silicon - the current release and the one from this PR.